framework/src/headers/ip/v6/mod.rs

```diff
@@ -0,0 +1,312 @@
+
+impl EndOffset for Ipv6Header {
+    // Note this does not allow IPv6-in-IPv6 or any other tunneling variants.
+    type PreviousHeader = MacHeader;
+
+    #[inline]
+    fn offset(&self) -> usize {
+        // IPv6 Header is always 40 bytes: (4 + 8 + 20 + 16 + 8 + 8 + 128 + 128) / 8 = 40
+        40
+    }
+
+    #[inline]
+    fn size() -> usize {
+        // Struct is always 40 bytes as well
+        40
+    }
+
+    #[inline]
+    fn payload_size(&self, _: usize) -> usize {
+        self.payload_len() as usize
+    }
+
+    #[inline]
+    fn check_correct(&self, _prev: &MacHeader) -> bool {
+        // prev.etype() == 0x86DD
+        true
+    }
+}
+
+// IPv6 Extension headers (and the first header) all have the "next header"
+// field which specifies the contents of the following extension or protocol
+// header. The value of this field is defined by IANA:
+// https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
+//
+// Similarly to being generic over v4 and v6 headers as parents for TCP/UDP, we
+// make extension headers generic but specify that they need to expose an
+// accessor for the next header field.
+pub trait Ipv6VarHeader: EndOffset + Default {
+    fn next_header(&self) -> u8;
+}
+
+// The IPv6 protocol header has a next header field and can be the
+// `PreviousHeader` type for any extension header.
+impl Ipv6VarHeader for Ipv6Header {
+    fn next_header(&self) -> u8 {
+        self.next_header()
+    }
+}
+
+// Formats the header for printing
+impl fmt::Display for Ipv6Header {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let src = Ipv6Addr::from(self.src());
+        let dst = Ipv6Addr::from(self.dst());
+        write!(
+            f,
+            "{} > {} version: {} traffic_class: {} flow_label: {} len: {} next_header: {} hop_limit: {}",
+            src,
+            dst,
+            self.version(),
+            self.traffic_class(),
+            self.flow_label(),
+            self.payload_len(),
+            self.next_header(),
+            self.hop_limit()
+        )
+    }
+}
+
+impl Ipv6Header {
+    #[inline]
+    pub fn flow(&self) -> Option<FlowV6> {
+        let next_hdr = self.next_header();
+        let src_ip = self.src();
+        let dst_ip = self.dst();
+        if (next_hdr == 6 || next_hdr == 17) && self.payload_size(0) >= 4 {
+            unsafe {
+                let self_as_u8 = (self as *const Ipv6Header) as *const u8;
+                let port_as_u8 = self_as_u8.offset(self.offset() as isize);
+                let port_slice = slice::from_raw_parts(port_as_u8, 4);
+                let dst_port = BigEndian::read_u16(&port_slice[..2]);
+                let src_port = BigEndian::read_u16(&port_slice[2..]);
+                Some(FlowV6 {
+                    src_ip: src_ip,
+                    dst_ip: dst_ip,
+                    src_port: src_port,
+                    dst_port: dst_port,
+                    proto: next_hdr,
```

Comment:
Note that with SRH added, this will no longer point to the L4 protocol. Need to figure out how to skip the extension headers. Same for the ports as well.

Reply:
👍 really good point. Can you look into that @seancribbs.

Reply:
Let's discuss Monday, this does not look straightforward.